src/pool/pool_disjoint_internal.h

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#ifndef UMF_POOL_DISJOINT_INTERNAL_H
+#define UMF_POOL_DISJOINT_INTERNAL_H 1
+
+#include <assert.h>
+#include <ctype.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <umf/memory_pool.h>
+#include <umf/memory_pool_ops.h>
+#include <umf/memory_provider.h>
+#include <umf/pools/pool_disjoint.h>
+
+#include "critnib/critnib.h"
+#include "uthash/utlist.h"
+
+#include "base_alloc_global.h"
+#include "provider/provider_tracking.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+#include "utils_log.h"
+#include "utils_math.h"
+#include "utils_sanitizers.h"
+
+typedef struct bucket_t bucket_t;
+typedef struct slab_t slab_t;
+typedef struct slab_list_item_t slab_list_item_t;
+typedef struct disjoint_pool_t disjoint_pool_t;
+
+typedef struct bucket_t {
+    size_t size;
+
+    // Linked list of slabs which have at least 1 available chunk.
+    // We always count available slabs as an optimization.
+    slab_list_item_t *available_slabs;
+    size_t available_slabs_num;
+
+    // Linked list of slabs with 0 available chunk.
+    slab_list_item_t *unavailable_slabs;
+
+    // Protects the bucket and all the corresponding slabs
+    utils_mutex_t bucket_lock;
+
+    // Reference to the allocator context, used access memory allocation
+    // routines, slab map and etc.
+    disjoint_pool_t *pool;
+
+    umf_disjoint_pool_shared_limits_handle_t shared_limits;
+
+    // For buckets used in chunked mode, a counter of slabs in the pool.
+    // For allocations that use an entire slab each, the entries in the Available
+    // list are entries in the pool.Each slab is available for a new
+    // allocation.The size of the Available list is the size of the pool.
+    // For allocations that use slabs in chunked mode, slabs will be in the
+    // Available list if any one or more of their chunks is free.The entire slab
+    // is not necessarily free, just some chunks in the slab are free. To
+    // implement pooling we will allow one slab in the Available list to be
+    // entirely empty. Normally such a slab would have been freed. But
```

Comment:
I think it's better to describe what API does, not what it doesn't do, to avoid confusion. `Normally(..)` sentence could be dropped. Same below before `Instead(...)`.